src/lune/builtins/datetime/mod.rs

```diff
@@ -0,0 +1,222 @@
+        });
+
+        methods.add_method(
+            "formatTime",
+            |_, this, (timezone, fmt_str, locale): (LuaValue, LuaString, LuaString)| {
+                Ok(this
+                    .format_time(
+                        Timezone::from_lua(timezone, &Lua::new())?,
+                        fmt_str.to_string_lossy(),
+                        locale.to_string_lossy(),
+                    )
+                    .map_err(|()| LuaError::external("failed to parse DateTime object, invalid")))
+            },
+        );
+
+        methods.add_method("toUniversalTime", |_, this: &DateTime, ()| {
+            Ok(this.to_universal_time())
+        });
+
+        methods.add_method("toLocalTime", |_, this: &DateTime, ()| {
+            Ok(this.to_local_time())
+        });
+    }
+}
+
+impl<'lua> FromLua<'lua> for DateTime {
+    fn from_lua(value: LuaValue<'lua>, _: &'lua Lua) -> LuaResult<Self> {
+        match value {
+            LuaValue::Nil => panic!("found nil"),
+            LuaValue::Table(t) => Ok(DateTime::from_unix_timestamp(
+                TimestampType::Seconds,
+                t.get("unixTimestamp")?,
+            )),
+            _ => panic!("invalid type"),
+        }
+    }
+}
+
+impl LuaUserData for DateTimeBuilder {
+    fn add_fields<'lua, F: LuaUserDataFields<'lua, Self>>(fields: &mut F) {
+        fields.add_field_method_get("year", |_, this| Ok(this.year));
+        fields.add_field_method_get("month", |_, this| Ok(this.month));
+        fields.add_field_method_get("day", |_, this| Ok(this.day));
+        fields.add_field_method_get("hour", |_, this| Ok(this.hour));
+        fields.add_field_method_get("minute", |_, this| Ok(this.minute));
+        fields.add_field_method_get("second", |_, this| Ok(this.second));
+        fields.add_field_method_get("millisecond", |_, this| Ok(this.millisecond));
+    }
+}
+
+impl<'lua> FromLua<'lua> for DateTimeBuilder {
+    fn from_lua(value: LuaValue<'lua>, _: &'lua Lua) -> LuaResult<Self> {
+        match value {
+            LuaValue::Table(t) => Ok(Self::default()
+                .with_year(t.get("year")?)
+                .with_month(
+                    (match t.get("month")? {
+                        LuaValue::String(str) => Ok(str.to_str()?.parse::<Month>().or(Err(
+                            LuaError::external("could not cast month string to Month"),
+                        ))?),
+                        LuaValue::Nil => {
+                            Err(LuaError::external("cannot find mandatory month argument"))
+                        }
+                        LuaValue::Number(num) => Ok(Month::try_from(num as u8).or(Err(
+                            LuaError::external("could not cast month number to Month"),
+                        ))?),
+                        LuaValue::Integer(int) => Ok(Month::try_from(int as u8).or(Err(
+                            LuaError::external("could not cast month integer to Month"),
```

Comment:
These type casts (using `as`) will panic at runtime if the user passes an invalid arg, maybe we could also split month parsing out into a separate function to make this `FromLua` implementation more concise?

Reply:
`as` cannot panic at runtime!

Reply:
You're right, I was remembering some old case where float -> int conversions using `as` would fail, seems like that doesn't happen anymore either.